### PR TITLE
Enable stricter MyPy typing for Rohmu

### PR DESCRIPTION
**Pull request by giacomo-alzetta-aiven**
# About this change - What it does

This PR enables various strict MyPy settings and adds/changes a lot of type hints to enable better typing support.

Some improvements can be added later on. For example, I/O typing is particularly messy in general. The current best practice would be to define a couple `Protocol`s and use that as arguments, why leaving the return types as concrete `io` classes. Unfortunately this is not simple because some stubs/library use the `typing.IO*` stuff which is not 100% compatible. Most Rohmu entities deal with `BufferedIOBase`, and so I took that as most general I/O thing, with a couple of `cast`s or `Union` in a few places where the `typing.IO` is inevitable.

Resolves: #13 

# Why this way

The API has not changed, except for the better types, but I expect that if you get a `TypeError` with the new version you probably have a bug. In a few places some plain attributes are now `@property`s, so they can't be set anymore.

### CODE REVIEW

File: rohmu/object_storage/google.py
Comment (fingon) [inline rohmu/object_storage/google.py:462]: No need for this return value - I think google implementations store_file_object returning something is a bug, not a feature (others do not return anything and in general API should be what is in base.py).

File: rohmu/object_storage/google.py
Comment (fingon) [inline rohmu/object_storage/google.py:543]: about all other backends return None, so should this.

File: rohmu/typing.py
Comment (fingon) [inline rohmu/typing.py:49]:
```suggestion
    def compress(self, data: bytes) -> bytes:
```

**Review by fingon — CHANGES_REQUESTED** [review]
Overall good work. Some comments.

In general what 'most' of object storages do should be API in base.py, and then all object storages should strictly implement that particular API (in terms of signature). If there's divergence, it is most likely bug and not feature.

File: rohmu/encryptor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/encryptor.py:26]: Yeah. rohmu.spec currently has `python3-cryptography >= 0.8` which is from 2015, version 1.6 is from the end of 2016 so I don't think many people are using current rohmu with that old cryptography versions.

File: rohmu/common/statsd.py
Comment (giacomo-alzetta-aiven) [reply rohmu/common/statsd.py:97]: This one does not work:

```
python3 -m mypy --show-error-codes rohmu/ test/
rohmu/common/statsd.py:101: error: Argument 1 to "update" of "MutableMapping" has incompatible type "Dict[str, Union[int, str, None]]"; expected "SupportsKeysAndGetItem[str, str]"  [arg-type]
rohmu/common/statsd.py:102: error: Argument "tags" to "increase" of "StatsClient" has incompatible type "Dict[str, str]"; expected "Optional[Dict[str, Union[int, str, None]]]"  [arg-type]
rohmu/common/statsd.py:102: note: "Dict" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
rohmu/common/statsd.py:102: note: Consider using "Mapping" instead, which is covariant in the value type
```

Sumit defined the `Tags` as `Dict[str, Union[int, str, None]]`. I have no idea where these tags come from and what values can they have but here they can be `str`, `int` or `None`. I can change the `Any` with `Union[str, int, None]` though.

File: rohmu/atomic_opener.py
Comment (giacomo-alzetta-aiven) [reply rohmu/atomic_opener.py:65]: Yes, without this running on Python3.7 fails with:
```
ImportError: cannot import name 'Literal' from 'typing' (/usr/lib64/python3.7/typing.py)
```
(e.g. executing the unittests).

But I think this can be simplified a lot. I believe the current implementation was done to avoid using any type comments... but IMHO if we can use 4 lines instead of 20 with a type comment and when moving to python3.8 simply remove the workaround I think it's better.

File: rohmu/object_storage/base.py
Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/base.py:199]: @fingon  The problem here is that `iter_key` returns a sequence of `IterKeyItem` which is defined as:

```
IterKeyItem = namedtuple("IterKeyItem", ["type", "value"])
```

so the `value` attribute is `Any` . Maybe we should redefine this to be generic in some way so we can use the type variable to provide a better result type? Or can you tell me if the `value` is actually always fixed?

Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/base.py:199]: Okay, so we could use this definition:

```
from typing import TypeVar, Generic, NamedTuple
Value = TypeVar("Value")


class IterKeyItem(NamedTuple, Generic[Value]):
    type: str
    value: Value
```

But then all uses of `IterKeyItem` will need a type argument . I guess the `Value` type depends on the provider?

Comment (fingon) [reply rohmu/object_storage/base.py:199]: no, value is str and type is enum (either TYPE_PREFIX or TYPE_OBJECT)

Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/base.py:199]: @fingon  Are you sure? 

In base.py inside `BaseTransfer.delete_tree` this is what I find:
```
names = [item["name"] for item in self.list_path(key, with_metadata=False, deep=True)]
```

and `list_path` is just a wrapper for `list_iter`:
```
    def list_path(self, key: str, *, with_metadata: bool = True, deep: bool = False) -> list[Any]:
        return list(self.list_iter(key, with_metadata=with_metadata, deep=deep))
```

Which yields the `value`s from `iter_key`:

```
    def list_iter(self, key: str, *, with_metadata: bool = True, deep: bool = False) -> Iterator[Any]:
        for item in self.iter_key(key, with_metadata=with_metadata, deep=deep):
            if item.type == KEY_TYPE_OBJECT:
                yield item.value
```

So `value` can't always be `str`, otherwise the `item["name"]` in delete tree would fail.

Or am I missing something?

Comment (fingon) [reply rohmu/object_storage/base.py:199]: Ah, right, the value's type depends on the type - PREFIX is str and OBJECT is indeed (platform dependant) dict. Even for those, we should really define fixed TypedDict but perhaps not worth it yet.

File: rohmu/compressor.py
Comment (fingon) [inline rohmu/compressor.py:28]:
```suggestion
def CompressionFile(dst_fp: HasRead, algorithm: str, level: int = 0, threads: int = 0) -> io.BufferedIOBase:
```

File: rohmu/compressor.py
Comment (fingon) [inline rohmu/compressor.py:67]:
```suggestion
def DecompressionFile(src_fp: HasRead, algorithm: str) -> io.BufferedIOBase:
```

File: rohmu/encryptor.py
Comment (fingon) [inline rohmu/encryptor.py:128]:
```suggestion
    def __init__(self, src_fp: HasRead, rsa_public_key_pem: Union[str, bytes]) -> None:
```

File: rohmu/encryptor.py
Comment (fingon) [inline rohmu/encryptor.py:247]: nit: this is really weird classmethod and not your problem but it could be really something outside class too.

File: rohmu/compressor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/compressor.py:67]: Yeah, changes this actually requires more changes downstream to FileWrap & all the various subclasses. `read` is not enough for them.

I believe we use at least: `read`, `write`, `flush`, `close` and `fileno`, possibly other methods in that hierarchy so I guess we need an other protocol for that other than `HasRead`.

I'm going to try and see what we end up with.

**Review by fingon — COMMENTED** [review]
Looks good to me now (+- unhappiness with formatting so not approving it yet). I'd like to get second opinion for this though.

**Comment by giacomo-alzetta-aiven** [comment]
I rebased the PR to avoid conflicts in rohmufile.py.

File: mypy.ini
Comment (rominf) [inline mypy.ini:2]: I believe we should use `python_version = 3.7` here, since it's the lowest supported version for the project. To make typing more powerful and annotations easier, you can add an additional dependency: https://pypi.org/project/typing-extensions/3.10.0.2/ (this exact version, to be bumped to https://pypi.org/project/typing-extensions/4.4.0/ in near future).

Comment (giacomo-alzetta-aiven) [reply mypy.ini:2]: It probably makes sense to change this python 3.9 to either 3.7 or 3.10. Not sure what @fingon  thinks about this. 

About the suggestion of checking python 3.11 discussed in the daily today, apparently the snappy package does not yet support 3.11 and from this issue it seems likely that they might never support it since the current maintainer is suggesting moving to cramjam... So moving Rohmu to python3.11 might have that issue.

If we need to change this I would prefer moving to 3.10 since 3.7 is EOL in ~1.5 months, and we should probably make sure that the unittest we at least import all modules (to catch runtime errors due to annotation definitions).
I will add the typing extensions module in the requirements explicitly.

Comment (giacomo-alzetta-aiven) [reply mypy.ini:2]: Okay, the typing_extensions package is already provided because it is required by a number of other packages:
```
$ pip show typing_extensions
Name: typing_extensions
Version: 4.5.0
Summary: Backported and Experimental Type Hints for Python 3.7+
Home-page: 
Author: 
Author-email: "Guido van Rossum, Jukka Lehtosalo, Łukasz Langa, Michael Lee" <levkivskyi@gmail.com>
License: 
Location: /home/giacomo.alzetta/Projects/rohmu/venv/lib/python3.10/site-packages
Requires: 
Required-by: azure-core, azure-storage-blob, google-api-python-client-stubs, mypy, mypy-boto3-sqs, pydantic
```
I believe it makes sense to include it explicitly. I would directly use version 4.5.0 which still supports python3.7.

Comment (fingon) [reply mypy.ini:2]: 3.7 is really soon EOL (c.f. https://endoflife.date/python) so I wouldn't go for that at least, 3.8 might make sense though.

File: rohmu/encryptor.py
Comment (bugant) [inline rohmu/encryptor.py:44]: nit: Wouldn't be better to raise a proper exception here instead of using `assert`? (like you did in https://github.com/aiven/rohmu/pull/104/files#diff-5c190a4f634c25382cbcfbc9997d46e5668951c55b395c4082beb4c7e3077312R92)

File: rohmu/encryptor.py
Comment (bugant) [inline rohmu/encryptor.py:168]: nit: Also here, I'd use a proper execption within an `if`

File: rohmu/object_storage/azure.py
Comment (bugant) [inline rohmu/object_storage/azure.py:187]: nit: Is it possible to cast to `Metadata` instead of `Dict[str, Any]`?

**Review by bugant — APPROVED** [review]
It looks good to me.  I left a couple of minor (non-blocking) comments.

File: mypy.ini
Comment (giacomo-alzetta-aiven) [reply mypy.ini:2]: IMHO, python3.7 is getting EOLed so I wouldn't want to move back there and next month having to change this again. For now I'm leaving the weird setup. In a next step we should probably move to python3.10.

For Python3.11 there is an issue with the snappy dependency which will need to be resolved...

File: rohmu/object_storage/azure.py
Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/azure.py:187]: Yes, but here the cast is on the "wrapper" object from which we extract the metadata.

Technically `Metadata` is just `Dict[str, Any]` but since I'm using it to refer to whatever is found under the `metadata` key I want to avoid using it outside that context. In the future the `Metadata` might become some `TypedDict` and suddenly it would not work correctly as used here.

File: rohmu/encryptor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/encryptor.py]: I decided to go with a custom `UninitializedException`.

File: rohmu/object_storage/azure.py
Comment (sjamgade) [inline rohmu/object_storage/azure.py:80]: this is not entirely correct,
proxy_info has `port` which should be of type `int`

use `ProxyInfo` instead ?

**Review by sjamgade — COMMENTED** [review]
noticed some loose assumptions. It would be nice to have them corrected in this or a follow up PR.

Havent looked at stuff under delta/* but rest LGTM

File: rohmu/notifier/http.py
Comment (giacomo-alzetta-aiven) [reply rohmu/notifier/http.py:135]: Yes, you are right. I just checked and it seems all places that call this pass through `sanitize_metadata`. I will fix this soon.

File: rohmu/__init__.py
Comment (akudiyar) [inline rohmu/__init__.py:20]: Have you considered using "tagged unions" for typing these transfer objects? 
See https://docs.pydantic.dev/latest/usage/types/#discriminated-unions-aka-tagged-unions
It seems better than casting to a "base" type

File: rohmu/atomic_opener.py
Comment (akudiyar) [inline rohmu/atomic_opener.py:40]: q: Why do we have to ignore that, what was the error?

File: rohmu/delta/common.py
Comment (akudiyar) [inline rohmu/delta/common.py:84]: `Dict`?

File: rohmu/inotify.py
Comment (akudiyar) [inline rohmu/inotify.py:61]: `Dict`? Or maybe `Mapping` is better since it's an inbound partameter

File: rohmu/inotify.py
Comment (akudiyar) [inline rohmu/inotify.py:51]: Some docstring with output explanation or better a named tuple could be useful here

File: rohmu/notifier/http.py
Comment (akudiyar) [inline rohmu/notifier/http.py:90]: Will `from __future__ import annotations` help with not using a string hint here?

File: rohmu/object_storage/azure.py
Comment (akudiyar) [inline rohmu/object_storage/azure.py:74]: It seems that we need named parameters in this constructor, because most of them are optional

File: rohmu/notifier/null.py
Comment (akudiyar) [inline rohmu/notifier/null.py]: `Dict`?

File: rohmu/notifier/logger.py
Comment (akudiyar) [inline rohmu/notifier/logger.py]: `Dict`?

File: rohmu/object_storage/azure.py
Comment (akudiyar) [inline rohmu/object_storage/azure.py:178]: It seems better to check the type here instead of casting, isn't it?

**Review by akudiyar — CHANGES_REQUESTED** [review]
Not a full review because the PR is refreshing often, but still some questions

**Review by akudiyar — DISMISSED** [review]
Not a full review because the PR is refreshing often, but still some questions

File: rohmu/atomic_opener.py
Comment (giacomo-alzetta-aiven) [reply rohmu/atomic_opener.py:40]: @akudiyar I believe this is something I had added to make mypy work with 3.7 setting as suggested by Roman, but since in the end I decided to just keep 3.9 and let a future PR decide which version we should target I guess this could be removed together with a few other comments.

File: rohmu/delta/common.py
Comment (giacomo-alzetta-aiven) [reply rohmu/delta/common.py:84]: There is a mix of `Dict` vs `dict` here but it's mostly due to python 3.7 support where `dict` does not support arguments.

Since I added the `from __future__ import annotations` flag in a few flag there is no issue with using `dict` with annotations.

There are a couple of places where we *must* use `Dict` (e.g. defining a type alias).

I guess I should replace all uses of `Dict` in annotations with `dict` and only leave `Dict` for type aliases until we drop python3.7 support.

File: rohmu/notifier/http.py
Comment (giacomo-alzetta-aiven) [reply rohmu/notifier/http.py:90]: Probably. I certainly added this before adding the future import. So I will review all these cases after lunch.

File: rohmu/object_storage/azure.py
Comment (giacomo-alzetta-aiven) [reply rohmu/object_storage/azure.py:178]: IMHO since you cannot do `isinstance(..., Dict[str, str])` checking that `item.value` is a `dict` plus that the value associated to key `name` is a `str` is going to be more complicated.

File: rohmu/__init__.py
Comment (giacomo-alzetta-aiven) [reply rohmu/__init__.py:20]: That seems like work for a different PR IMHO... Here I just fixed the type to use stricter mypy settings which was the goal of this PR. What you suggest seems to require some changes in the class hierarchy to make it work.

File: rohmu/object_storage/azure.py
Comment (fingon) [reply rohmu/object_storage/azure.py:74]: Public API -> please no signature change unless really needed.

File: rohmu/compressor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/compressor.py:27]: The level values depends on the compression algorithm. Most have a parameter from 0 to 9, ZSTD has a range 1-22. But, IMHO, I don't think that using `Literal[int]` here would give us anything. It would only make harder to write code that uses this function without much benefits. 

The `Literal[str]` in the `atomic_opener` makes much more sense because the `@overload` can be used to specify `TextIO` vs `BinaryIO` in return type, but whether the compression level is `1`, `100` or 1 million doesn't really change much and seems only a configuration parameter, so I'd rather leave this as plain `int`.

File: rohmu/delta/snapshot.py
Comment (giacomo-alzetta-aiven) [reply rohmu/delta/snapshot.py:54]: They are available from python3.9. Using deferred annotations you can import the code even in python3.7, even though it has limitations (e.g. cannot use this in Pydantic model fields).  We are running mypy for python3.9 atm.

IMHO after merging this we can easily add a PR to just replace all uses of List/Tuple etc with list/tuple where possible. I don't really want to delay this PR further for this kind of change.

File: rohmu/inotify.py
Comment (giacomo-alzetta-aiven) [reply rohmu/inotify.py:72]: That would be the only use of Mapping in rohmu. For this PR I'd prefer using the built-in, if later we want to use Mapping and other more generic types we can go hunt all the places where we could change those and do so.

File: rohmu/object_storage/azure.py
Comment (fingon) [inline rohmu/object_storage/azure.py:188]: nit perhaps, but this is more readable for me at least:
```suggestion
        for item in items:
            # We expect single result but Azure listing is prefix match so we need to explicitly
            # look up the matching result
            item_name: str
            if item.type == KEY_TYPE_OBJECT:
                assert isinstance(item.value, dict)
                item_name = item.value["name"]
            else:
                assert isinstance(item.value, str)
                item_name = item.value
            if item_name.rstrip("/").rsplit("/", 1)[-1] == expected_name:
                break
        else:
            raise FileNotFoundFromStorageError(path)
        if item.type != KEY_TYPE_OBJECT:
            raise FileNotFoundFromStorageError(path)
```

**Review by fingon — CHANGES_REQUESTED** [review]
Sorry, still perhaps one or two worth changing (and rest nit-ish).

File: rohmu/encryptor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/encryptor.py]: Fixed

File: rohmu/encryptor.py
Comment (giacomo-alzetta-aiven) [reply rohmu/encryptor.py]: You are right. In the other places the cipher/authenticator etc were already public. Fixed this.

**Review by fingon — APPROVED** [review]

**Pull request merged**
